tortoise/fixture_test.go

```diff
@@ -0,0 +1,791 @@
+		for _, o := range opt.opts {
+			o(&header)
+		}
+	}
+	val := &atxAction{
+		reg:     s.reg,
+		header:  header,
+		ballots: map[uint32]*ballotAction{},
+	}
+	s.reg.register(val)
+	s.atxs[epoch] = val
+	return val
+}
+
+type atxAction struct {
+	reg registry
+
+	header    types.AtxTortoiseData
+	ballots   map[uint32]*ballotAction
+	reference *ballotAction
+}
+
+func (a *atxAction) String() string {
+	return fmt.Sprintf("atx %v", a.header.ID)
+}
+
+func (*atxAction) deps() []action {
+	return nil
+}
+
+type ballotOpt func(*ballotAction)
+
+type bopt struct {
+	opts []ballotOpt
+}
+
+func (b *bopt) beacon(value string) *bopt {
+	b.opts = append(b.opts, func(ballot *ballotAction) {
+		if ballot.EpochData == nil {
+			ballot.EpochData = &types.ReferenceData{}
+		}
+		copy(ballot.EpochData.Beacon[:], value)
+	})
+	return b
+}
+
+func (b *bopt) eligibilities(value int) *bopt {
+	b.opts = append(b.opts, func(ballot *ballotAction) {
+		ballot.Eligibilities = uint32(value)
+	})
+	return b
+}
+
+func (b *bopt) activeset(values ...*atxAction) *bopt {
+	b.opts = append(b.opts, func(ballot *ballotAction) {
+		ballot.activeset = values
+		for _, val := range values {
+			if ballot.EpochData == nil {
+				ballot.EpochData = &types.ReferenceData{}
+			}
+			ballot.EpochData.ActiveSet = append(ballot.EpochData.ActiveSet, val.header.ID)
+		}
+	})
+	return b
+}
+
+// encoded votes.
+type evotes struct {
+	baseBallot *ballotAction
+	votes      types.Votes
+}
+
+func (e *evotes) base(base *ballotAction) *evotes {
+	e.votes.Base = base.ID
+	e.baseBallot = base
+	return e
+}
+
+func (e *evotes) abstain(lid uint32) *evotes {
+	e.votes.Abstain = append(e.votes.Abstain, types.GetEffectiveGenesis()+types.LayerID(lid))
+	return e
+}
+
+func (e *evotes) support(lid int, id string, height uint64) *evotes {
+	vote := types.Vote{
+		LayerID: types.GetEffectiveGenesis() + types.LayerID(lid),
+		Height:  height,
+	}
+	copy(vote.ID[:], id)
+	e.votes.Support = append(e.votes.Support, vote)
+	return e
+}
+
+func (e *evotes) against(lid uint32, id string, height uint64) *evotes {
+	vote := types.Vote{
+		LayerID: types.GetEffectiveGenesis().Add(lid),
+		Height:  height,
+	}
+	copy(vote.ID[:], id)
+	e.votes.Against = append(e.votes.Against, vote)
+	return e
+}
+
+func (b *bopt) votes(value *evotes) *bopt {
+	b.opts = append(b.opts, func(ballot *ballotAction) {
+		ballot.base = value.baseBallot
+		ballot.Opinion.Votes = value.votes
+	})
+	return b
+}
+
+func (a *atxAction) execute(trt *Tortoise) {
+	trt.OnAtx(&a.header)
+}
+
+func (a *atxAction) ballot(n int, opts ...*bopt) *ballotAction {
+	lid := uint32(n) + types.GetEffectiveGenesis().Uint32()
+	if val, exist := a.ballots[lid]; exist {
+		return val
+	}
+	b := types.BallotTortoiseData{}
+	b.AtxID = a.header.ID
+	b.Layer = types.LayerID(lid)
+	hs := hash.Sum(a.header.ID[:], []byte(strconv.Itoa(int(lid))))
+	copy(b.ID[:], hs[:])
+
+	val := &ballotAction{
+		BallotTortoiseData: b,
+	}
+	for _, opt := range opts {
```

Comment:
options relating to `EpochData` is incompatible if a.reference != nil and should be set conditionally to be more realistic?

Reply:
i will set Ref only if EpochData is nil